crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+#[macro_use]
+extern crate serde;
+
+use hyperx::header::Header;
+use oauth2::{Client as OauthClient, StandardToken};
+use serde::Deserialize;
+use std::convert::TryFrom;
+use url::Url;
+use www_authenticate::{Challenge, ChallengeFields, RawChallenge, WwwAuthenticate};
+
+use crate::errors::*;
+use crate::reference::Reference;
+
+const OCI_VERSION_KEY: &str = "Docker-Distribution-Api-Version";
+
+pub mod errors;
+pub mod reference;
+
+type OciResult<T> = Result<T, anyhow::Error>;
+
+struct Client {
+    token: Option<StandardToken>,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Client { token: None }
+    }
+}
+
+impl Client {
+    /// According to the v2 specification, 200 and 401 error codes MUST return the
+    /// version. It appears that any other response code should be deemed non-v2.
+    ///
+    /// For this implementation, it will return v2 or an error result. If the error is a
+    /// `reqwest` error, the request itself failed. All other error messages mean that
+    /// v2 is not supported.
+    pub async fn version(&self, host: String) -> OciResult<String> {
```

Comment:
Should `host` be `&str` for all these methods?

Reply:
Fixed.